Async validators + Recaptcha

omarryhan opened the pull request:
#13 

New features:
- Support for async and sync validators
- testing py37

The code I added isn't of best quality, especially the tests. This is only meant to be a prototype.

I think in order to merge this feature, we'd have to go 1 of the following paths:

1. Break backward compatibility and make `validate_on_submit` a coroutine.
2. Keep both `validate_on_submit` and `validate_on_submit_async` (maybe name it differently). However, if we do this, then the user shouldn't call `validate_on_submit_async` followed by `validate_on_submit`, because `validate_on_submit_async` will do lots of monkey patching (We can have some sort of flag though that explicitly enables this functionality).
3. Make a new `SanicForm` class. Maybe call it `SanicForm_` or sth
4. Do nothing :smile:

omarryhan committed:
async validate_on_submit prototype
now testing py37

omarryhan committed:
Improved tests structure
Added tests coverage

omarryhan committed:
More async validators tests

omarryhan committed:
Added support for async field.post and pre validators
Added pytest-asyncio in the list of dev dependencies
Added tests for wtforms patching

omarryhan committed:
Added Travis CI

omarryhan committed:
Fix travis 3.7

omarryhan committed:
Fix travis 3.7

omarryhan committed:
Sanic-wtf will now fail if a user patches wtforms and continues using the synchronous validate on submit method

omarryhan committed:
Sanic-wtf will now fail if a user patches wtforms and continues using the synchronous validate on submit method

omarryhan commented:
So, I did some cleanup. I think the only thing missing is patching FieldList. If left as is, then it will only be called synchronously by any given form. However, it will fail if any of it's methods are coroutines as they will not be awaited. Basically, as long as you don't use custom FieldList with coroutines in it, you're fine.

As for the tests coverage, I think I covered the main use cases. There might be some untested edge cases though, so if someone can help me spot some, it would be great.

omarryhan committed:
Pytest now warns if any
Minor fixes

omarryhan committed:
added doctstings

omarryhan committed:
Added support for Recaptcha
Updated readme
Bumped version to 0.7
100% backward compatible changes
More tests. Coverage at 86%
Docs still do not reflect the changes in the repo

omarryhan committed:
Documented Recaptcha integration

omarryhan committed:
100% recaptcha test cov

omarryhan committed:
Recaptcha docstrings

omarryhan committed:
Minor docstring fix

omarryhan committed:
Fixed typo

omarryhan committed:
Fixed typo

omarryhan committed:
Moved config_prefix into kwargs

omarryhan committed:
added test for config prefix

omarryhan committed:
Fixed more typos

omarryhan committed:
Removed unused patching code

omarryhan committed:
Added option for JS only

omarryhan committed:
Validators now validate in order. Cov @94%

pyx commented:
Thanks for the feature, I was very busy lately, I did not have time to go through the code right now, do you mind give me sometime.  Thanks again.

omarryhan commented:
Sure, np! :)

omarryhan committed:
Replaced recaptcha's validators param with extra_validaors

omarryhan committed:
recaptcha field passing list instead of set to super()

omarryhan committed:
Better recaptcha json/form error handling

omarryhan committed:
Recaptcha only using aiorecaptcha's error msg not the full error object

omarryhan committed:
Fixed test

omarryhan committed:
More efficient recaptcha validator

omarryhan committed:
Minor optimization tweak
Docstring fix

omarryhan committed:
Updated README
Better typing in Recaptcha field
More efficient recaptcha validator
Using isawaitable instead of iscoroutinefunction

omarryhan committed:
Removed commented code

pyx commented:
Hi Omar, sorry that I was super busy lately, it seems you put a lot of effort into this, I will go through this in a few days.

Thank you so much.

omarryhan committed:
Drop Python3.5 dependancy

omarryhan committed:
Hardcoded wtforms version to avoid breakage of patching logic on updates in wtforms

omarryhan committed:
Bumped version

omarryhan committed:
Remove python 3.5 testing from .travis yaml

omarryhan committed:
More descriptive code comments

pyx commented:
Hi Omar Ryhan
I am sorry I failed in manage time to update the project, are you still interested in using this? If you are interested, I will put you into  collaborators list.

pyx commented:
@mikekeda  How about this one, @omarryhan invested a lot of time into this, what's your thought?

pyx commented:
@mikekeda I at OP accidentally in last post.

omarryhan commented:
Hey @pyx. No worries.

I only passively maintain the Sanic projects I was working on. The Sanic team recently introduced some breaking changes to the request context API and I'm not really sure whether my PR might break things or not. My Sanic projects are still running on their latest LTS version which differs from their latest version. I'm pip installing from my own sanic-wtf repo, and things are working fine.

I suggest we leave this PR open until someone shows some interest and maybe tests things out on the latest Sanic version, then maybe we can merge it?

omarryhan commented:
I must also add that I monkey patched some of wtform's private methods to make this lib support async validations. The downside of this that wtforms cannot guarantee us that they won't break their private APIs. So I hard-coded the wtforms version to avoid things breaking. One downside of that is that we'll not be getting the latest security patches from wtforms should there be any.

pyx commented:
@omarryhan you are right, relying on private API is fragile to say the least and we should try to avoid. We will wait and see. 

Please remember to add your name in AUTHOR with next commit, not necessarily this one but can be something like a separate one updating python versions tested upon, you have the commit rights.
Thank you.

mikekeda commented on .travis.yml:
```diff
@@ -0,0 +1,15 @@
+install:
+  pip install tox-travis
+script: 
+  tox
```
add a new line to the end

mikekeda commented on sanic_wtf/__init__.py:
```diff
@@ -7,12 +7,16 @@
 
-__version__ = '0.6.0.dev0'
+from ._patch import patch
+from .recaptcha import RecaptchaField
```
it's better to use absolute imports

mikekeda commented on sanic_wtf/__init__.py:
```diff
@@ -110,7 +114,6 @@ def getlist(self, name, default=None):
         """
         return super().get(name, default)
 
-
```
according to pep8 here should be 2 blank lines, no need to remove 1 blank line
https://www.python.org/dev/peps/pep-0008/#blank-lines

mikekeda commented on sanic_wtf/__init__.py:
```diff
@@ -141,7 +149,35 @@ def __init__(self, request=None, *args, meta=None, **kwargs):
     def validate_on_submit(self):
+        ''' For async validators: use self.validate_on_submit_async.
+            This method is only still here for backward compatibility
+        '''
```
use double quotes for all docstring
https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+    self._errors = None
+    success = True
+    for name, field in self._fields.items():
+        if extra_validators is not None and name in extra_validators:
```
it's more pythonic to use `if extra_validators` instead of `extra_validators is not None`
but be aware that for example `if 0` returns False and `if 0 is not None` returns True

mikekeda commented on sanic_wtf/__init__.py:
```diff
@@ -7,12 +7,16 @@
 from wtforms.csrf.session import SessionCSRF
 from wtforms.meta import DefaultMeta
 from wtforms.validators import DataRequired, StopValidation
+from wtforms.fields.core import Field
```
Field isn't used

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+import functools
+import asyncio
+import warnings
+from wtforms import Form
```
those 3 imports aren't used

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+
+    # Added an await here
+    return await self.validate_base(extra)
+
```
add 1 blank line

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+    self.errors = list(self.process_errors)
+    stop_validation = False
+
+
```
remove blank line

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+        self.errors.append(e.args[0])
+
+    return len(self.errors) == 0
+
```
add 1 blank line

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+        except ValueError as e:  # Catches validation errors
+            self.errors.append(e.args[0])
+    return False    
+
```
add 1 blank line

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+            # # 3.2 Patch fieldlist type
+            # # Won't work asynchronously until "entries" (subfields) are dealt with
+        self.patched = True
+
```
add 1 blank line

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+        self.patched = True
+
+def patch(f):
+    ''' Patch wtforms to make it support async validators
```
use double quotes to docstrings

mikekeda commented on sanic_wtf/_patch.py:
```diff
@@ -0,0 +1,175 @@
+        # Await if a coroutine
+        res = self.pre_validate(form)
+        if inspect.isawaitable(res):
+            res = await(res)
```
res isn't used,
`await res` looks better to me compare to `await(res) `

mikekeda commented on sanic_wtf/recaptcha.py:
```diff
@@ -0,0 +1,239 @@
+
+
+__all__ = ['RecaptchaField']
+
```
add 1 blank line

mikekeda commented on sanic_wtf/recaptcha.py:
```diff
@@ -0,0 +1,239 @@
+        )
+    except aiorecaptcha.RecaptchaError as e:
+        raise ValidationError(e.msg)
+
```
add 1 blank line

mikekeda commented on sanic_wtf/recaptcha.py:
```diff
@@ -0,0 +1,239 @@
+        error_callback=field._config.get(self._config_prefix + '_ERROR_CALLBACK')
+    )
+    return Markup(js) + Markup(html)
+
```
add 1 blank line

mikekeda commented on sanic_wtf/recaptcha.py:
```diff
@@ -0,0 +1,239 @@
+    return Markup(js) + Markup(html)
+
+class RecaptchaField(Field):
+    '''
```
use double quotes for docstrings

mikekeda commented on tests/conftest.py:
```diff
@@ -15,3 +19,35 @@ async def add_session(request):
         request[name] = session
 
     return test_app
+
```
add 1 blank line before each fixture

mikekeda commented on tests/helpers.py:
```diff
@@ -0,0 +1,12 @@
+# taking shortcut here, assuming there will be only one "string" (the token)
+# ever longer than 40.
+csrf_token_pattern = '''value="([0-9a-f#]{40,})"'''
+
```
add 1 blank line

mikekeda commented on tests/test_async_validators.py:
```diff
@@ -0,0 +1,222 @@
+import re
+import asyncio
```
asyncio isn't used

mikekeda commented on tests/test_async_validators.py:
```diff
@@ -0,0 +1,222 @@
+import asyncio
+
+from sanic import response
+from wtforms.validators import DataRequired, Length, ValidationError
```
ValidationError isn't used

mikekeda commented on tests/test_async_validators.py:
```diff
@@ -0,0 +1,222 @@
+
+from sanic import response
+from wtforms.validators import DataRequired, Length, ValidationError
+from wtforms import FileField, StringField, SubmitField
```
FileField isn't used

mikekeda commented on tests/test_async_validators.py:
```diff
@@ -0,0 +1,222 @@
+from wtforms.validators import DataRequired, Length, ValidationError
+from wtforms import FileField, StringField, SubmitField
+
+from sanic_wtf import SanicForm, to_bytes
```
to_bytes isn't used

mikekeda commented on tests/test_async_validators.py:
```diff
@@ -0,0 +1,222 @@
+    req, resp = app.test_client.post('/', data=payload)
+    assert resp.status == 200
+    assert 'valid' in resp.text
+
```
add 1 blank line before each test

mikekeda commented on tests/test_form.py:
```diff
@@ -1,27 +1,14 @@
 # -*- coding: utf-8 -*-
 import re
 import os.path
+import asyncio
```
asyncio isn't used

mikekeda commented on tests/test_form.py:
```diff
@@ -1,27 +1,14 @@
 
 from sanic import response
-from wtforms.validators import DataRequired, Length
+from wtforms.validators import DataRequired, Length, ValidationError
```
ValidationError isn't used

mikekeda commented on tests/test_form.py:
```diff
@@ -173,7 +161,6 @@ async def index(request):
     assert resp.status == 200
     assert 'validated' in resp.text
 
-
```
why?
here should be 2 blank lines according to pep8 https://www.python.org/dev/peps/pep-0008/#blank-lines

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+import asyncio
+
+import pytest
+from wtforms import Form, Field, FieldList, StringField
```
Form, Field, FieldList aren't used

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+    3.1.2. wtforms.fields.core.Field._run_validation_chain() --> _run_validation_chain_async 
+    TODO: 3.2 wtforms.fields.core.FieldList
+'''
+
```
add 1 blank line before each function

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+# test validate methods
+def test_all_fields_have_async_validate_method():
+
+    async def async_val(form, field):
```
`form, field` aren't used, could we use `_, __` instead?

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+            ) is True
+
+# Pre/Post Validators
+
```
add 1 blank line

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+def req():
+    class App:
+        pass
+    class Request:
```
add 1 blank line before

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+@pytest.mark.asyncio
+async def test_post_validator_field_patch(req):
+    POST_VAL_ERR_MSG = 'post_patched_and_called'
+    class AsyncPostValidatorField(StringField):
```
add 1 blank line before

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+@pytest.mark.asyncio
+async def test_post_validator_field_sync(req):
+    POST_VAL_ERR_MSG = 'post_patched_and_called'
+    class AsyncPostValidatorField(StringField):
```
add 1 blank line before

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+        async def post_validate(self, form, stop_validation):
+            pass
+
+    async def fail(*args):
```
`*args` isn't used, could we use `*_` instead?

mikekeda commented on tests/test_patching.py:
```diff
@@ -0,0 +1,191 @@
+    assert await form.validate_on_submit_async() is False
+    assert 'I failed' in str(form.errors)
+    assert 'I failed again' in str(form.errors)
+  
```
remove spaces

mikekeda commented on tests/test_recaptcha.py:
```diff
@@ -0,0 +1,154 @@
+import asyncio
+
+import pytest
+from wtforms import ValidationError
```
ValidationError isn't used

mikekeda commented on tests/test_recaptcha.py:
```diff
@@ -0,0 +1,154 @@
+    _base_config = {'WTF_CSRF_ENABLED': False}
+    test_app = App({**_base_config, **config})
+    return Request(test_app)
+
```
add 1 blank line before each top level function
(here should be 2 blank lines in total - https://www.python.org/dev/peps/pep-0008/#blank-lines)

mikekeda commented on tests/test_recaptcha.py:
```diff
@@ -0,0 +1,154 @@
+        def __init__(self, config):
+            self.config = config
+
+
```
remove 1 blank line

mikekeda commented on sanic_wtf/recaptcha.py:
```diff
@@ -0,0 +1,239 @@
+    widget = recaptcha_widget
+            
+    def __init__(self, label='Recaptcha', extra_validators: Union[list, tuple, set]=None, **kwargs):
+        validators = set([recaptcha_validator])
```
`validators = {recaptcha_validator}`
will be more pythonic

mikekeda commented on sanic_wtf/recaptcha.py:
```diff
@@ -0,0 +1,239 @@
+
+    widget = recaptcha_widget
+            
+    def __init__(self, label='Recaptcha', extra_validators: Union[list, tuple, set]=None, **kwargs):
```
add spaces before and after `=` if you use typing
should be `extra_validators: Union[list, tuple, set] = None` instead of `extra_validators: Union[list, tuple, set]=None`

mikekeda commented on tox.ini:
```diff
@@ -1,6 +1,9 @@
 deps = -rrequirements-dev.txt
-commands = py.test
+commands = 
+    ./clear_pycache.sh 
```
why do we nee to clear pycache?

mikekeda commented:
> @mikekeda How about this one, @omarryhan invested a lot of time into this, what's your thought?

1. I think it's better to split this into 2 RPs Async validators and Recaptcha
2. monkey patching doesn't look good to me, can we create a separate class for this?
3. What performance improvement we will get with Async validators?
4. I don't think it's a good idea to make RECAPTCHA required, it's better to make it optional (in `RECAPTCHA_PUBLIC_KEY ` and `RECAPTCHA_PRIVATE_KEY ` are marked as required)
5. Why `clear` and `clear_pycache.sh` are committed?

omarryhan commented:
Thanks for taking the time to review @mikekeda 

I added async not for the performance gain but to be able to call async code for validations, like querying a database to check if an email exists.

I'll address these issues once people show enough interest in the PR. Right now, it seems like I'm the only one using it.
